### ISSUE TITLE
High memory usage when using .engine with YOLOv8 on CUDA

### ISSUE THREAD

**Issue by Nicqhhq** [ultralytics/ultralytics#22275]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am experiencing very high memory usage when loading a model exported as TensorRT (.engine) in Ultralytics YOLOv8, even for a small model (yolo11n.engine).
	•	Hardware:
	•	GPU: RTX 4060
	•	CPU: i7-12700F
	•	RAM: 64 GB
	•	CUDA: 13.0
	•	Ultralytics YOLO: latest version

Issue:
	1.	Using .pt (PyTorch) with .to("cpu") consumes ~700 MB of RAM.
	2.	Using .engine (TensorRT) consumes ~1.4 GB of RAM, even without explicitly moving it to CUDA.

code example 

`from ultralytics import YOLO
import cv2

# TensorRT engine model
model_path = "./models/yolo11n.engine"

# Load the model
model = YOLO(model_path, task="detect")

# Dummy frame
frame = cv2.imread("test_image.jpg")

# Inference on GPU
results = model.predict(source=frame, device=0, imgsz=640, conf=0.3, iou=0.2)`

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Nicqhhq, thank you for your interest in Ultralytics 🚀! This is an automated triage response — an Ultralytics engineer will join the thread soon to assist further.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For your TensorRT (.engine) memory usage report, please include:
-Exact ultralytics package version (`pip show ultralytics`)
-How the `.engine` was created (export command and args)
-Full inference snippet (minimal) and any runtime logs
-Environment details: OS, Python, CUDA, TensorRT versions, GPU driver, and `nvidia-smi` output
-A single test image and observed memory readings (RAM and VRAM)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks for the detailed hardware summary and the snippet. Once you share the MRE and environment specifics above, it will help us quickly reproduce and investigate the high memory usage you’re seeing with your `yolo11n.engine` setup ⚙️🧩

**Comment by glenn-jocher** [comment]
Thanks for the report—TensorRT `.engine` loads always create a CUDA context and pinned host buffers, so you’ll see higher host RAM vs a `.pt` on CPU; to minimize this, export a fixed-shape, small‑batch FP16 engine and match your inference size, which reduces TRT buffer allocations; see the TensorRT export arguments in the docs for details: [TensorRT export for YOLO11 models → Export arguments](https://docs.ultralytics.com/integrations/tensorrt/#export-arguments).

Try re‑exporting and testing with:
```bash
yolo export model=yolo11n.pt format=engine imgsz=640 batch=1 dynamic=False half=True nms=False
yolo predict model=yolo11n.engine source=test_image.jpg device=0 imgsz=640
```

To help us confirm there’s nothing unexpected in your environment, please share the text outputs of:
```bash
yolo checks
python -c "import tensorrt as trt; print(trt.__version__)"
nvidia-smi --query-gpu=memory.total,memory.used --format=csv,noheader
```
and note how you measured RAM vs VRAM; also confirm the args used to build `yolo11n.engine` (dynamic/batch/imgsz/precision).

**Comment by Nicqhhq** [comment]
I tested the suggested change, but unfortunately it did not reduce the host RAM usage. The consumption remains high.

Process memory usage (ps -o pid,comm,%mem,rss -C python -C python3):
Old run:

`2444 python    2.2  1480532`

New run:

`60968 python3  2.0  1333284`


As you can see, the difference is minimal and the issue persists.


Environment checks:
$ yolo checks
Ultralytics 8.3.204 🚀 Python-3.10.12 torch-2.7.1+cu126 CUDA:0 (NVIDIA GeForce RTX 4060, 7925MiB)
Setup complete ✅ (20 CPUs, 62.6 GB RAM, 193.0/218.5 GB disk)

OS                     Linux-6.8.0-84-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.10.12
Install                pip
Path                   /home/openit/.local/lib/python3.10/site-packages/ultralytics
RAM                    62.63 GB
Disk                   193.0/218.5 GB
CPU                    12th Gen Intel Core(TM) i7-12700F
CPU count              20
GPU                    NVIDIA GeForce RTX 4060, 7925MiB
GPU count              1
CUDA                   12.6

numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.3>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 9.0.1>=7.1.2
pyyaml                 ✅ 5.4.1>=5.3.1
requests               ✅ 2.25.1>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.7.1>=1.8.0
torch                  ✅ 2.7.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.22.1>=0.9.0
psutil                 ✅ 7.0.0
polars                 ✅ 1.33.1
ultralytics-thop       ✅ 2.0.14>=2.0.0

$ python3 -c "import tensorrt as trt; print(trt.__version__)"
10.13.0.35

$ nvidia-smi --query-gpu=memory.total,memory.used --format=csv,noheader
8188 MiB, 529 MiB


So, in summary: the fixed-shape FP16 engine was exported and tested, but the RAM usage did not change in practice.

**Comment by glenn-jocher** [comment]
Thanks for checking—your RSS looks consistent with TensorRT 10.x: the runtime creates a CUDA context plus page‑locked host staging buffers sized by the optimization profile on first load, which shows up as ~1–2 GB RSS even at batch=1; Ultralytics doesn’t add extra host allocations beyond engine/context creation. To confirm it’s TRT (not Ultralytics), please compare with trtexec and watch RSS while it loads the same engine:
```bash
trtexec --loadEngine=./models/yolo11n.engine --duration=1
```
If host RAM is a hard constraint, ONNX Runtime CUDA usually holds a smaller host footprint:
```bash
yolo export model=yolo11n.pt format=onnx
yolo predict model=yolo11n.onnx device=0 imgsz=640
```
For implementation details, see the exporter path that builds the engine and how workspace limits are applied in the TensorRT builder config in our references: [export_engine method](https://docs.ultralytics.com/reference/engine/exporter/#export_engine) and [onnx2engine implementation](https://docs.ultralytics.com/reference/utils/export/__init__/#onnx2engine).
